test/test_save.py

```diff
@@ -0,0 +1,471 @@
+from pydicom.misc import is_dicom
+from pydicom.tag import ItemTag, SequenceDelimiterTag, Tag
+from pydicom.uid import UID, JPEGBaseline8Bit, generate_uid
+from wsidicom import WsiDicom
+from wsidicom.geometry import Point, Size, SizeMm
+from wsidicom.instance import ImageData, WsiDicomFile, WsiDicomFileWriter
+from wsidicom.uid import WSI_SOP_CLASS_UID
+from wsidicom.wsidicom import WsiDicomLevel
+
+wsidicom_test_data_dir = os.environ.get("WSIDICOM_TESTDIR", "C:/temp/wsidicom")
+sub_data_dir = "interface"
+data_dir = wsidicom_test_data_dir + '/' + sub_data_dir
+
+
+class WsiDicomTestFile(WsiDicomFile):
+    """Test version of WsiDicomFile that overrides __init__."""
+    def __init__(self, filepath: Path, transfer_syntax: UID, frame_count: int):
+        self._filepath = filepath
+        self._fp = DicomFile(filepath, mode='rb')
+        self._fp.is_little_endian = transfer_syntax.is_little_endian
+        self._fp.is_implicit_VR = transfer_syntax.is_implicit_VR
+        self._frame_count = frame_count
+        self.__enter__()
+
+
+class WsiDicomTestImageData(ImageData):
+    def __init__(self, data: Sequence[bytes], tiled_size: Size) -> None:
+        if len(data) != tiled_size.area:
+            raise ValueError('Number of frames and tiled size area differ')
+        TILE_SIZE = Size(10, 10)
+        self._data = data
+        self._tile_size = TILE_SIZE
+        self._image_size = tiled_size * TILE_SIZE
+
+    @property
+    def files(self) -> List[Path]:
+        return []
+
+    @property
+    def transfer_syntax(self) -> UID:
+        return JPEGBaseline8Bit
+
+    @property
+    def image_size(self) -> Size:
+        return self._image_size
+
+    @property
+    def tile_size(self) -> Size:
+        return self._tile_size
+
+    @property
+    def pixel_spacing(self) -> SizeMm:
+        return SizeMm(1.0, 1.0)
+
+    @property
+    def samples_per_pixel(self) -> int:
+        return 3
+
+    @property
+    def photometric_interpretation(self) -> str:
+        return 'YBR'
+
+    def _get_decoded_tile(
+        self,
+        tile_point:
+        Point,
+        z: float,
+        path: str
+    ) -> Image.Image:
+        raise NotImplementedError
+
+    def _get_encoded_tile(self, tile: Point, z: float, path: str) -> bytes:
+        return self._data[tile.x + tile.y * self.tiled_size.width]
+
+    def close(self) -> None:
+        return super().close()
+
+
+@pytest.mark.save
+class WsiDicomFileSaveTests(unittest.TestCase):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.test_folders: Dict[Path, WsiDicom]
+
+    @classmethod
+    def setUpClass(cls):
+        cls.tiled_size = Size(10, 10)
+        cls.frame_count = cls.tiled_size.area
+        MIN_FRAME_LENGTH = 2
+        MAX_FRAME_LENGTH = 100
+        # Generate test data by itemizing random bytes of random length
+        # from MIN_FRAME_LENGTH to MAX_FRAME_LENGTH.
+        cls.test_data = [
+            urandom(randint(
```

Comment:
You typically want your tests to be repeatable in case they fail. So I would remove the randomness here either by using a seed or something else.